File: dekaf-core/src/main/java/org/jetbrains/jdba/sql/Scriptum.java
package org.jetbrains.jdba.sql;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import org.jetbrains.jdba.core.ResultLayout;

import java.util.ArrayList;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

import static org.jetbrains.jdba.util.Strings.eq;
import static org.jetbrains.jdba.util.Strings.rtrim;



/**
 * Scriptum provides with access to SQL texts that are placed in separate files
 * (not within Java code).
 *
 * @author Leonid Bushuev from JetBrains
 */
public final class Scriptum {

  @NotNull
  private final ScriptumResource[] myResources;

  @Nullable
  private final String myDialect;



  //// CONSTRUCTORS \\\\

  @NotNull
  public static Scriptum of(@NotNull final Class clazz) {
    ClassLoader classLoader = clazz.getClassLoader();
    String className = clazz.getName();
    String path = className.replace('.', '/');
    String resourceName = path + ".sql";

    return of(classLoader, resourceName);
  }

  @NotNull
  public static Scriptum of(@NotNull final Class clazz, @NotNull final String name) {
    ClassLoader classLoader = clazz.getClassLoader();
    String packageName = clazz.getPackage().getName();
    String path = packageName.replace('.', '/');
    String fileName = name;
    if (!fileName.contains(".")) fileName += ".sql";
    String resourceName = path + '/' + fileName;

    return of(classLoader, resourceName);
  }

  @NotNull
  public static Scriptum dialectOf(@NotNull final Class clazz, @Nullable final String dialect) {
    ClassLoader classLoader = clazz.getClassLoader();
    String className = clazz.getName();
    String path = className.replace('.', '/');
    ArrayList<ScriptumResource> sr = new ArrayList<ScriptumResource>(2);

    if (dialect != null) {
      String name1 = path + '+' + dialect + ".sql";
      boolean exist1 = classLoader.getResource(name1) != null;
      if (exist1) {
        ScriptumResourceFromJava r1 = new ScriptumResourceFromJava(classLoader, name1);
        sr.add(r1);
      }
    }

    String name2 = path + ".sql";
    boolean exist2 = classLoader.getResource(name2) != null;
    if (exist2) {
      ScriptumResourceFromJava r2 = new ScriptumResourceFromJava(classLoader, name2);
      sr.add(r2);
    }

    if (sr.isEmpty())
      throw new IllegalArgumentException(String.format("Resources for class %s not found",
                                                       className));

    ScriptumResource[] resources = sr.toArray(new ScriptumResource[sr.size()]);
    return new Scriptum(resources, dialect);
  }

  @NotNull
  public static Scriptum dialectOf(@NotNull final Scriptum parentScriptum,
                                   @Nullable final String dialect) {
    if (eq(parentScriptum.myDialect, dialect)) {
      return parentScriptum;
    }
    else {
      return new Scriptum(parentScriptum.myResources, dialect);
    }
  }


  @NotNull
  public static Scriptum of(final ClassLoader classLoader, final String resourceName) {
    ArrayList<ScriptumResource> sr = new ArrayList<ScriptumResource>(2);
    boolean exists = classLoader.getResource(resourceName) != null;
    if (exists) {
      ScriptumResourceFromJava r2 = new ScriptumResourceFromJava(classLoader, resourceName);
      sr.add(r2);
    }
    else {
      throw new IllegalArgumentException(String.format("Resources for class %s not found",
                                                       resourceName));
    }

    ScriptumResource[] resources = sr.toArray(new ScriptumResource[sr.size()]);
    return new Scriptum(resources, null);
  }



  private Scriptum(@NotNull final ScriptumResource[] resources,
                   @Nullable final String dialect) {
    myResources = resources;
    myDialect = dialect;
  }


  //// FUNCTIONS TO QUERY TEXTS \\\\

  /**
   * Finds the requested script.
   *
   * <p>
   *   If no such script, returns null.
   * </p>
   *
   * @param name  name of the script to find.
   * @return      the found script, or <i>null</i> if not found.
   */
  @Nullable
  public final TextFileFragment findText(@NotNull String name) {
    TextFileFragment fragment;
    final String nameWithDialect = myDialect == null ? null : name + '+' + myDialect;
    for (int i = myResources.length-1; i >= 0; i--) {
      ScriptumResource r = myResources[i];
      if (nameWithDialect != null) {
        fragment = r.find(nameWithDialect);
        if (fragment != null) return fragment;
      }
      fragment = r.find(name);
      if (fragment != null) return fragment;
    }
    return null;
  }

  /**
   * Provides with the requested script.
   *
   * <p>
   *   If no such script, throws {@link ScriptNotFoundException}.
   * </p>
   *
   * @param name  name of the script to provide with.
   * @return      the script.
   * @throws ScriptNotFoundException if no such script.
   */
  @NotNull
  public final TextFileFragment getText(@NotNull String name) throws ScriptNotFoundException{
    TextFileFragment fragment = findText(name);
    if (fragment != null) {
      return fragment;
    }
    else {
      // some useful diagnostics
      StringBuilder b = new StringBuilder();
      b.append("No such script with name: ").append(name).append('\n');
      boolean was = false;
      for (ScriptumResource r : myResources) {
        for (String existentName : r.getExistentNames()) {
          if (!was) {
            b.append("There are scripts: ");
            was = true;
          }
          else {
            b.append(", ");
          }
          b.append(existentName);
        }
      }
      if (!was) {
        b.append("There are no scripts at all");
      }
      throw new ScriptNotFoundException(b.toString());
    }
  }


  @NotNull
  public final <S> SqlQuery<S> query(@NotNull final String name,
                                     @NotNull final ResultLayout<S> layout) {
    TextFileFragment fragment = getText(name);
    fragment = stripSingleStatement(fragment);
    final SqlQuery<S> query = new SqlQuery<S>(fragment, layout);
    query.setDisplayName(fragment.getFragmentName());
    return query;
  }

  @NotNull
  public final SqlCommand command(@NotNull final String name) {
    TextFileFragment fragment = getText(name);
    fragment = stripSingleStatement(fragment);
    return new SqlCommand(fragment);
  }

  @NotNull
  public final SqlScript script(@NotNull final String name) {
    TextFragment fragment = getText(name);
    SqlScriptBuilder b = new SqlScriptBuilder();
    b.parse(fragment.text);
    return new SqlScript(b.build());
  }


  private static final Pattern STRIP_SINGLE_STATEMENT_PATTERN =
      Pattern.compile("((\\;\\s*)+|(\\n\\/\\s*\\n?)+)$", Pattern.DOTALL);

  @NotNull
  private TextFileFragment stripSingleStatement(@NotNull final TextFileFragment fragment) {
    Matcher m = STRIP_SINGLE_STATEMENT_PATTERN.matcher(fragment.text);
    if (m.find()) {
      int n = fragment.text.length();
      n -= m.group(1).length();
      final String text = rtrim(fragment.text.substring(0, n));
      return new TextFileFragment(text,
                                  fragment.getTextName(),
                                  fragment.row, fragment.pos,
                                  fragment.getFragmentName());
    }
    else {
      return fragment;
    }
  }



  //// EXCEPTIONS \\\\

  public static class ScriptNotFoundException extends RuntimeException {
    private ScriptNotFoundException(final String message) {
      super(message);
    }
  }

}

File: DekafMain/src/text/Scriptum.kt
package org.jetbrains.dekaf.text

import org.jetbrains.dekaf.core.QueryResultLayout
import org.jetbrains.dekaf.sql.SqlCommand
import org.jetbrains.dekaf.sql.SqlQuery
import org.jetbrains.dekaf.sql.SqlScript
import org.jetbrains.dekaf.sql.SqlScriptBuilder
import org.jetbrains.dekaf.util.*

import java.util.ArrayList
import java.util.regex.Pattern



/**
 * Scriptum provides with access to SQL texts that are placed in separate files
 * (not within Java code).

 * @author Leonid Bushuev from JetBrains
 */
class Scriptum
private constructor
(
        private val myResources: Array<ScriptumResource>,

        private val myDialect: String?
)
{

    ////// INITIALIZATION \\\\\\

    companion object {

        fun of(clazz: Class<*>): Scriptum {
            val classLoader = clazz.classLoader
            val className = clazz.name
            val path = className.replace('.', '/')
            val resourceName = path + ".sql"

            return of(classLoader, resourceName)
        }

        fun of(clazz: Class<*>, name: String): Scriptum {
            val classLoader = clazz.classLoader
            val packageName = clazz.`package`.name
            val path = packageName.replace('.', '/')
            var fileName = name
            if (!fileName.contains(".")) fileName += ".sql"
            val resourceName = path + '/' + fileName

            return of(classLoader, resourceName)
        }

        fun dialectOf(clazz: Class<*>, dialect: String?): Scriptum {
            val classLoader = clazz.classLoader
            val className = clazz.name
            val path = className.replace('.', '/')
            val sr = ArrayList<ScriptumResource>(2)

            if (dialect != null) {
                val name1 = "$path+$dialect.sql"
                val exist1 = classLoader.getResource(name1) != null
                if (exist1) {
                    val r1 = ScriptumResourceFromJava(classLoader, name1)
                    sr.add(r1)
                }
            }

            val name2 = path + ".sql"
            val exist2 = classLoader.getResource(name2) != null
            if (exist2) {
                val r2 = ScriptumResourceFromJava(classLoader, name2)
                sr.add(r2)
            }

            if (sr.isEmpty())
                throw IllegalArgumentException("Resources for class $className not found")

            val resources = sr.toTypedArray()
            return Scriptum(resources, dialect)
        }

        fun dialectOf(parentScriptum: Scriptum, dialect: String?): Scriptum {
            if (parentScriptum.myDialect == dialect) {
                return parentScriptum
            }
            else {
                return Scriptum(parentScriptum.myResources, dialect)
            }
        }


        fun of(classLoader: ClassLoader, resourceName: String): Scriptum {
            val sr = ArrayList<ScriptumResource>(2)
            val exists = classLoader.getResource(resourceName) != null
            if (exists) {
                val r2 = ScriptumResourceFromJava(classLoader, resourceName)
                sr.add(r2)
            }
            else {
                throw IllegalArgumentException(String.format("Resources for class %s not found",
                                                             resourceName))
            }

            val resources = sr.toTypedArray()
            return Scriptum(resources, null)
        }

        private val STRIP_SINGLE_STATEMENT_PATTERN =
                Pattern.compile("""((;\s*)+|(\n/\s*\n?)+)$""",
                                Pattern.DOTALL)
    }



    ////// FUNCTIONS TO QUERY TEXTS \\\\\\

    /**
     * Finds the requested script.
     *
     * If no such script, returns null.
     *
     * @param name  name of the script to find.
     * @return      the found script, or *null* if not found.
     */
    fun findText(name: String): TextFileFragment? {
        var fragment: TextFileFragment?
        val nameWithDialect = if (myDialect == null) null else name + '+' + myDialect
        for (i in myResources.indices.reversed()) {
            val r = myResources[i]
            if (nameWithDialect != null) {
                fragment = r.find(nameWithDialect)
                if (fragment != null) return fragment
            }
            fragment = r.find(name)
            if (fragment != null) return fragment
        }
        return null
    }

    /**
     * Provides with the requested script.
     *
     * If no such script, throws [ScriptNotFoundException].
     *
     * @param name  name of the script to provide with.
     * @return      the script.
     * @throws ScriptNotFoundException if no such script.
     */
    @Throws(ScriptNotFoundException::class)
    fun getText(name: String): TextFileFragment {
        val fragment = findText(name)
        if (fragment != null) {
            return fragment
        }
        else {
            // some useful diagnostics
            val b = StringBuilder()
            b.append("No such script with name: ").append(name).append('\n')
            var was = false
            for (r in myResources) {
                for (existentName in r.existentNames) {
                    if (!was) {
                        b.append("There are scripts: ")
                        was = true
                    }
                    else {
                        b.append(", ")
                    }
                    b.append(existentName)
                }
            }
            if (!was) {
                b.append("There are no scripts at all")
            }
            throw ScriptNotFoundException(b.toString())
        }
    }


    fun <S> query(name: String,
                  layout: QueryResultLayout<S>): SqlQuery<S> {
        var fragment = getText(name)
        fragment = stripSingleStatement(fragment)
        val query = SqlQuery(fragment, layout)
        query.displayName = fragment.fragmentName
        return query
    }

    fun command(name: String): SqlCommand {
        var fragment = getText(name)
        fragment = stripSingleStatement(fragment)
        return SqlCommand(fragment)
    }

    fun script(name: String): SqlScript {
        val fragment = getText(name)
        val b = SqlScriptBuilder()
        b.parse(fragment.text)
        return SqlScript(b.build())
    }

    private fun stripSingleStatement(fragment: TextFileFragment): TextFileFragment {
        val m = STRIP_SINGLE_STATEMENT_PATTERN.matcher(fragment.text)
        if (m.find()) {
            var n = fragment.text.length
            n -= m.group(1).length
            val text = fragment.text.substring(0, n).rtrim()
            return TextFileFragment(text,
                                    fragment.textName,
                                    fragment.row, fragment.pos,
                                    fragment.fragmentName)
        }
        else {
            return fragment
        }
    }


    ////// EXCEPTIONS \\\\\\

    class ScriptNotFoundException internal constructor(message: String) : RuntimeException(message)

}
